### PR TITLE
Add test cases to ensure coverage with exited validators

### PR DESCRIPTION
**Pull request by ralexstokes**
The tests in #2560 uncovered some missing coverage in the pyspec test suite.

This PR adds test cases for spots in the spec where differences between the active and exited validator sets could lead to difference in state transitions.

Tests added:
- `test_random_with_exits_with_duplicates`
	- randomize state to ensure exited validators are present, then check a block with a sync aggregate with randomized participation
- `test_random_with_exits_without_duplicates`
	- see previous, but without duplicated members in the sync committee
- `test_balance_threshold_with_exited_validators`
	- “Exercise a very specific failure mode where exited validators are incorrectly included in the total active balance when weighing justification.”
- `test_slashings_with_random_state`
	- test effects of (collective) slashing processing with randomized state including some exited validators
- `test_full_random_4`
	- test rewards with a randomized state containing some exited (but not slashed) validators
- `test_full_random_without_leak_0`
	- test rewards with a randomized state containing some exited (but not slashed) validators that is _not_ leaking

### CODE REVIEW

File: tests/core/pyspec/eth2spec/test/helpers/rewards.py [hunk 1]
```diff
@@ -255,7 +255,19 @@ def run_get_inactivity_penalty_deltas(spec, state):
                 else:
                     assert penalties[index] > base_penalty
         else:
-            assert penalties[index] == 0
+            if not is_post_altair(spec):
```
Comment (ralexstokes) [inline tests/core/pyspec/eth2spec/test/helpers/rewards.py:258]: comment: this function is getting a bit unwieldy...

File: tests/core/pyspec/eth2spec/test/helpers/sync_committee.py [hunk 2]
```diff
@@ -161,23 +163,3 @@ def _build_block_for_next_slot_with_sync_participation(spec, state, committee_in
...
-    randao_index = (current_epoch + 1) % spec.EPOCHS_PER_HISTORICAL_VECTOR
-    while True:
-        committee = spec.get_next_sync_committee_indices(state)
-        if duplicates:
-            if len(committee) != len(set(committee)):
-                return committee
-        else:
-            if len(committee) == len(set(committee)):
-                return committee
-        state.randao_mixes[randao_index] = hash(state.randao_mixes[randao_index])
```
Comment (ralexstokes) [inline tests/core/pyspec/eth2spec/test/helpers/sync_committee.py:183]: this has been entirely removed for the better

File: tests/core/pyspec/eth2spec/test/altair/block_processing/sync_aggregate/test_process_sync_aggregate_random.py [hunk 3]
```diff
@@ -18,8 +27,8 @@
 )
 
 
-def _test_harness_for_randomized_test_case(spec, state, duplicates=False, participation_fn=None):
-    committee_indices = get_committee_indices(spec, state, duplicates=duplicates)
+def _test_harness_for_randomized_test_case(spec, state, expect_duplicates=False, participation_fn=None):
+    committee_indices = compute_committee_indices(spec, state, state.current_sync_committee)
```
Comment (ralexstokes) [inline tests/core/pyspec/eth2spec/test/altair/block_processing/sync_aggregate/test_process_sync_aggregate_random.py:31]:
```suggestion
    committee_indices = compute_committee_indices(spec, state)
```

File: tests/core/pyspec/eth2spec/test/helpers/random.py [hunk 4]
```diff
@@ -128,3 +128,33 @@ def randomize_state(spec, state, rng=Random(8020), exit_fraction=None, slash_fra
     exit_random_validators(spec, state, rng, fraction=exit_fraction)
     slash_random_validators(spec, state, rng, fraction=slash_fraction)
     randomize_attestation_participation(spec, state, rng)
+
+
+def patch_state_to_non_leaking(spec, state):
```
Comment (ralexstokes) [inline tests/core/pyspec/eth2spec/test/helpers/random.py:133]: re-org this function into `helpers` as it is used in multiple locations now

File: tests/core/pyspec/eth2spec/test/helpers/random.py [hunk 5]
```diff
@@ -128,3 +128,33 @@ def randomize_state(spec, state, rng=Random(8020), exit_fraction=None, slash_fra
...
+    This function performs an irregular state transition so that:
+    1. the current justified checkpoint references the previous epoch
+    2. the previous justified checkpoint references the epoch before previous
+    3. the finalized checkpoint matches the previous justified checkpoint
+
+    The effects of this function are intended to offset randomization side effects
+    performed by other functionality in this module so that if the ``state`` was leaking,
+    then the ``state`` is not leaking after.
+    """
+    state.justification_bits = (True, True, True, True)
```
Comment (ralexstokes) [inline tests/core/pyspec/eth2spec/test/helpers/random.py:144]:
```suggestion
    state.justification_bits[0] = True
    state.justification_bits[1] = True
```

Comment (ralexstokes) [reply tests/core/pyspec/eth2spec/test/helpers/random.py:144]: going to commit this -- if something strange starts blowing up, check here

File: tests/core/pyspec/eth2spec/test/helpers/sync_committee.py [hunk 6]
```diff
@@ -153,6 +154,7 @@ def _build_block_for_next_slot_with_sync_participation(spec, state, committee_in
             state,
             block.slot - 1,
             [index for index, bit in zip(committee_indices, committee_bits) if bit],
+            block_root=block.parent_root,
```
Comment (ralexstokes) [inline tests/core/pyspec/eth2spec/test/helpers/sync_committee.py:157]: passing in block root if we have it, to avoid accidentally re-computing via an entirely new scratch block

File: tests/core/pyspec/eth2spec/test/phase0/epoch_processing/test_process_justification_and_finalization.py [hunk 7]
```diff
@@ -300,3 +302,75 @@ def test_12_ok_support_messed_target(spec, state):
...
+
+    # Step 1: Exit ~1/2 vals in current epoch
+    epoch = spec.get_current_epoch(state)
+    for index in spec.get_active_validator_indices(state, epoch):
+        if rng.choice([True, False]):
+            continue
+
+        validator = state.validators[index]
+        validator.exit_epoch = epoch
+        validator.withdrawable_epoch = epoch + 1
```
Comment (djrtwo) [inline tests/core/pyspec/eth2spec/test/phase0/epoch_processing/test_process_justification_and_finalization.py:332]: `+ 1` is unrealistic. Suggest using the min withdrawability constant. Might save us trouble when writing withdrawal tests in the future

File: tests/core/pyspec/eth2spec/test/phase0/epoch_processing/test_process_justification_and_finalization.py [hunk 8]
```diff
@@ -300,3 +302,75 @@ def test_12_ok_support_messed_target(spec, state):
 @spec_state_test
 def test_12_poor_support(spec, state):
     yield from finalize_on_12(spec, state, 3, False, False)
+
+
+@with_all_phases
+@spec_state_test
+def test_balance_threshold_with_exited_validators(spec, state):
```
Comment (djrtwo) [inline tests/core/pyspec/eth2spec/test/phase0/epoch_processing/test_process_justification_and_finalization.py:309]: Great test!

**Review by djrtwo — COMMENTED** [review]
Awesome! Some very minor comments before merge. 

Can you also add a short list of the tests in the pr comment so if we look back at this a year later, it'll be clear what the meat was here

File: tests/core/pyspec/eth2spec/test/helpers/random.py (hunk 4 above)
Comment (ralexstokes) [reply tests/core/pyspec/eth2spec/test/helpers/random.py:133]: correct, just consolidation

File: tests/core/pyspec/eth2spec/test/altair/block_processing/sync_aggregate/test_process_sync_aggregate_random.py [hunk 9]
```diff
@@ -98,7 +107,25 @@ def test_random_misc_balances_and_half_participation_with_duplicates(spec, state
...
+        expect_duplicates=True,
+        participation_fn=lambda comm: rng.sample(comm, len(comm) // 2),
+    )
+
+
+@with_altair_and_later
+@with_presets([MAINNET], reason="to create duplicate committee")
+@spec_state_test
+@single_phase
+def test_random_with_exits_with_duplicates(spec, state):
```
[the remark this answers is not in the document]
Comment (ralexstokes) [reply tests/core/pyspec/eth2spec/test/altair/block_processing/sync_aggregate/test_process_sync_aggregate_random.py:119]: will put in separate PR, there may be a few other edge cases to get in here

**Review by djrtwo — APPROVED** [review]
great! left the two comments about future changes "unresolved"
